### PR TITLE
Adjust ghost-measured-template rotation logic

### PR DESCRIPTION
**Pull request by FolkvangrForgent**
Adjust ghost-measured-template placed from chat to behave more like placed templates with regards to rotation. Added additional logic for hex grids. This should complement pr #4143 and pr #4173 which addressed issue #4126 .

- Set ctrl to rotate by 5deg (this did not change)
- Set ctrl+shift to rotate by 15deg (this used to be 15deg for square or grid-less and 30deg otherwise)
- Set shift to rotate by 60deg on hex and 45deg on anything else (this used to just be 45deg which was quite useless on hex)
- Compensate for hex row format by offsetting the original angle by 30deg so 60deg aligns properly. (this could be omitted if the hex angle is set to 30deg but makes hex much more usable)
- Reduce recalculations. (set this.document.direction to value just calculated in this.document._source.direction instead of recalculating how much the template should be rotated)

### CODE REVIEW

File: src/module/canvas/ghost-measured-template.ts [hunk 1]
```diff
@@ -13,6 +13,13 @@ export class GhostTemplate extends MeasuredTemplatePF2e {
         const snapped = canvas.grid.getSnappedPosition(center.x, center.y, 2);
         this.document.x = snapped.x;
         this.document.y = snapped.y;
+        if (
+            this.moveTime === 0 &&
+            (canvas.grid.type === CONST.GRID_TYPES.HEXODDR || canvas.grid.type === CONST.GRID_TYPES.HEXEVENR)
+        ) {
+            this.document._source.direction += 30;
+            this.document.direction = this.document._source.direction;
```
Comment (stwlam) [inline src/module/canvas/ghost-measured-template.ts:21]: Use `this.document.updateSource` instead of assigning twice

**Review by stwlam — CHANGES_REQUESTED** [review]

File: src/module/canvas/ghost-measured-template.ts (hunk 1 above)
Comment (FolkvangrForgent) [reply src/module/canvas/ghost-measured-template.ts:21]: I am unaware of how this.document.updateSource(...) works as the original code did not use and there doesn't seem to be alot of examples. Would the following be correct?
```
            this.document.direction += 30;
            this.document.updateSource({ direction: this.document.direction });
```

Comment (stwlam) [reply src/module/canvas/ghost-measured-template.ts:21]: If you're adding to the value, you'd want:
```ts
this.document.updateSource({ direction: this.document.direction + 30 });
```
That will trigger data repreparation and set both.

**Comment by FolkvangrForgent** [comment]
Switch to updateSource() function instead of double setting values.
Switched directly setting position or rotation to use updateSource().
Removed double snapping to Grid on left click as source position should now be grid aligned.

File: src/module/canvas/ghost-measured-template.ts [hunk 2]
```diff
@@ -11,17 +11,18 @@ export class GhostTemplate extends MeasuredTemplatePF2e {
...
         const snapped = canvas.grid.getSnappedPosition(center.x, center.y, 2);
-        this.document.x = snapped.x;
-        this.document.y = snapped.y;
+        if (
+            this.moveTime === 0 &&
+            (canvas.grid.type === CONST.GRID_TYPES.HEXODDR || canvas.grid.type === CONST.GRID_TYPES.HEXEVENR)
+        ) {
+            this.document.updateSource({ direction: this.document.direction + 30 });
+        }
+        this.document.updateSource({ direction: this.document.direction, x: snapped.x, y: snapped.y });
```
Comment (stwlam) [inline src/module/canvas/ghost-measured-template.ts:20]:
```suggestion
        if (
            this.moveTime === 0 &&
            (canvas.grid.type === CONST.GRID_TYPES.HEXODDR || canvas.grid.type === CONST.GRID_TYPES.HEXEVENR)
        ) {
            this.document.updateSource({ direction: this.document.direction + 30 });
        }
        this.document.updateSource({ x: snapped.x, y: snapped.y });
```

File: src/module/canvas/ghost-measured-template.ts [hunk 3]
```diff
@@ -32,16 +33,22 @@ export class GhostTemplate extends MeasuredTemplatePF2e {
...
             event.stopPropagation();
-            const delta = canvas.grid.type > CONST.GRID_TYPES.SQUARE ? 30 : 15;
-            const snap = event.shiftKey ? delta : 5;
-            this.document._source.direction += snap * Math.sign(event.deltaY);
-            this.document.direction += snap * Math.sign(event.deltaY);
+            const snap = event.shiftKey ? 15 : 5;
+            this.document.updateSource({
+                direction: this.document.direction + snap * Math.sign(event.deltaY),
+                x: this.document.x,
+                y: this.document.y,
```
Comment (stwlam) [inline src/module/canvas/ghost-measured-template.ts:40]: Why does this need to be updated?

File: src/module/canvas/ghost-measured-template.ts [hunk 4]
```diff
@@ -32,16 +33,22 @@ export class GhostTemplate extends MeasuredTemplatePF2e {
...
             event.stopPropagation();
-            const snap = 45;
-            this.document._source.direction += snap * Math.sign(event.deltaY);
-            this.document.direction += snap * Math.sign(event.deltaY);
+            const snap =
+                canvas.grid.type >= CONST.GRID_TYPES.HEXODDR && canvas.grid.type <= CONST.GRID_TYPES.HEXEVENQ ? 60 : 45;
+            this.document.updateSource({
+                direction: this.document.direction + snap * Math.sign(event.deltaY),
+                x: this.document.x,
+                y: this.document.y,
```
Comment (stwlam) [inline src/module/canvas/ghost-measured-template.ts:50]: Ditto

**Review by stwlam — CHANGES_REQUESTED** [review]

File: src/module/canvas/ghost-measured-template.ts (hunk 3 above)
Comment (FolkvangrForgent) [reply src/module/canvas/ghost-measured-template.ts:40]: Oops made an assumption about behavior that was not correct, I'll remove the x and y setting.

File: src/module/canvas/ghost-measured-template.ts (hunk 4 above)
Comment (FolkvangrForgent) [reply src/module/canvas/ghost-measured-template.ts:50]: ditto

**Comment by FolkvangrForgent** [comment]
Removed extra variable setting in updateSource() from incorrect assumption about how updateSource() works.

File: src/module/canvas/ghost-measured-template.ts [hunk 5]
```diff
@@ -11,17 +11,18 @@ export class GhostTemplate extends MeasuredTemplatePF2e {
...
         const snapped = canvas.grid.getSnappedPosition(center.x, center.y, 2);
-        this.document.x = snapped.x;
-        this.document.y = snapped.y;
+        if (
+            this.moveTime === 0 &&
+            (canvas.grid.type === CONST.GRID_TYPES.HEXODDR || canvas.grid.type === CONST.GRID_TYPES.HEXEVENR)
+        ) {
+            this.document.updateSource({ direction: this.document.direction + 30 });
+        }
+        this.document.updateSource({ x: snapped.x, y: snapped.y });
```
Comment (stwlam) [inline src/module/canvas/ghost-measured-template.ts:20]:
```suggestion
        const hexTypes: number[] = [CONST.GRID_TYPES.HEXODDR, CONST.GRID_TYPES.HEXEVENR];
        const direction =
            this.moveTime === 0 && hexTypes.includes(canvas.grid.type)
                ? this.document.direction + 30
                : this.document.direction;
        this.document.updateSource({ x: snapped.x, y: snapped.y, direction });

```

File: src/module/canvas/ghost-measured-template.ts [hunk 6]
```diff
@@ -11,17 +11,18 @@ export class GhostTemplate extends MeasuredTemplatePF2e {
...
         const snapped = canvas.grid.getSnappedPosition(center.x, center.y, 2);
-        this.document.x = snapped.x;
-        this.document.y = snapped.y;
+        const hexTypes: number[] = [CONST.GRID_TYPES.HEXODDR, CONST.GRID_TYPES.HEXEVENR];
+        const direction =
+            this.moveTime === 0 && hexTypes.includes(canvas.grid.type)
+                ? this.document.direction + 30
+                : this.document.direction;
+        this.document.updateSource({ x: snapped.x, y: snapped.y, direction });
+
```
Comment (stwlam) [inline src/module/canvas/ghost-measured-template.ts:20]:
```suggestion
        const hexTypes: number[] = [CONST.GRID_TYPES.HEXODDR, CONST.GRID_TYPES.HEXEVENR];
        const direction =
            this.moveTime === 0 && hexTypes.includes(canvas.grid.type)
                ? this.document.direction + 30
                : this.document.direction;
        this.document.updateSource({ ...snapped, direction });

```

**Review by stwlam — APPROVED** [review]

**Pull request merged**